tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/SymbStateRewriterImplWithArrays.scala

```diff
@@ -3,7 +3,120 @@ package at.forsyte.apalache.tla.bmcmt
+          -> List(new SetCupRule(this)), // TODO: consider copying one array and storing the edges of the other
+        key(tla.filter(tla.name("x"), tla.name("S"), tla.name("p")))
+          -> List(new SetFilterRule(this)),
+        key(tla.map(tla.name("e"), tla.name("x"), tla.name("S")))
+          -> List(new SetMapRule(this)),
+        key(OperEx(ApalacheOper.expand, tla.name("X")))
+          -> List(new SetExpandRule(this)),
+        key(tla.powSet(tla.name("X")))
+          -> List(new PowSetCtorRule(this)),
+        key(tla.union(tla.enumSet()))
+          -> List(new SetUnionRule(this)),
+        key(tla.dotdot(tla.int(1), tla.int(10)))
+          -> List(new IntDotDotRule(this, intRangeCache)),
+        // integers
+        key(tla.lt(tla.int(1), tla.int(2)))
+          -> List(new IntCmpRule(this)),
+        key(tla.le(tla.int(1), tla.int(2)))
+          -> List(new IntCmpRule(this)),
+        key(tla.gt(tla.int(1), tla.int(2)))
+          -> List(new IntCmpRule(this)),
+        key(tla.ge(tla.int(1), tla.int(2)))
+          -> List(new IntCmpRule(this)),
+        key(tla.plus(tla.int(1), tla.int(2)))
+          -> List(new IntArithRule(this)),
+        key(tla.minus(tla.int(1), tla.int(2)))
+          -> List(new IntArithRule(this)),
+        key(tla.mult(tla.int(1), tla.int(2)))
+          -> List(new IntArithRule(this)),
+        key(tla.div(tla.int(1), tla.int(2)))
+          -> List(new IntArithRule(this)),
+        key(tla.mod(tla.int(1), tla.int(2)))
+          -> List(new IntArithRule(this)),
+        key(tla.exp(tla.int(2), tla.int(3)))
+          -> List(new IntArithRule(this)),
+        key(tla.uminus(tla.int(1)))
+          -> List(new IntArithRule(this)),
+        // -----------------------------------------------------------------
+        // RULES BELOW WERE ADDED TO RUN UNIT TESTS, WILL BE LOOKED AT LATER
+        // -----------------------------------------------------------------
+        // logic
+        key(OperEx(ApalacheOper.skolem, tla.exists(tla.name("x"), tla.name("S"), tla.name("p"))))
+          -> List(new QuantRule(this)),
+        key(tla.exists(tla.name("x"), tla.name("S"), tla.name("p")))
+          -> List(new QuantRule(this)),
+        key(tla.forall(tla.name("x"), tla.name("S"), tla.name("p")))
+          -> List(new QuantRule(this)),
+        // control flow
+        key(tla.ite(tla.name("cond"), tla.name("then"), tla.name("else")))
+          -> List(new IfThenElseRule(this)),
+        // functions
+        key(tla.funDef(tla.name("e"), tla.name("x"), tla.name("S")))
+          -> List(new FunCtorRule(this)),
+        key(tla.appFun(tla.name("f"), tla.name("x")))
+          -> List(new FunAppRule(this)),
+        // tuples, records, and sequences
+        key(tla.tuple(tla.name("x"), tla.int(2)))
+          -> List(new TupleOrSeqCtorRule(this)),
+        key(tla.enumFun(tla.str("a"), tla.int(2)))
+          -> List(new RecCtorRule(this))
+    )
```

Comment:
The end goal is to do the overwriting as you described, once all rules are accounted for. I've done this because I wanted to only use a subset of all rules for now.

Reply:
In what scenario would you want to delete an existing rule and not replace it with anything (= not copying it over in this implementation)?

Reply:
The current one, in which we are still developing the encoding. Basically I don't want to add rules to the encoding before going over them, even if they remain unchanged. Getting a "noRuleApplicable" exception (I think it is written like this) is what I want for the rules we haven't gone over yet.

Reply:
In that case, I'd still use copy and then manually delete the rules you don't want AND write a comment above explaining how this is temporary.

Reply:
I see, I will try to implement it as you suggested.

Reply:
I'm trying to override `ruleLookupTable` with `(super.ruleLookupTable ++ newRules) -- unsupportedRules.keys`, but end up getting `super may not be used on <none>`. Is there a correct way to access the map in the parent class?

Reply:
Scala handles inheritance of `val` in an [unexpected way,](https://docs.scala-lang.org/tutorials/FAQ/initialization-order.html) at least to me. My fix was to make a `defaultRuleLookupTable` that could be accessed by the child class. If there is a better way, please let me know.

Reply:
This solution looks  fine to me.